go/storage/mkvs/checkpoint/checkpoint_test.go

```diff
@@ -491,3 +492,337 @@ func testPruneGapAfterCheckpointRestore(t *testing.T, factory dbApi.Factory) {
+	hash1 := cp1.EncodedHash()
+	hash2 := cp2.EncodedHash()
+
+	if !hash1.Equal(&hash2) {
+		return fmt.Errorf("checkpoint hash not equal: want %s, got %s", hash1, hash2)
+	}
+	return nil
+}
+
+// populateDB populates database with n random entries.
+func populateDB(ctx context.Context, ndb dbApi.NodeDB, ns common.Namespace, n uint16, rnd *rand.Rand) (node.Root, error) {
+	tree := mkvs.New(nil, ndb, node.RootTypeState)
+	defer tree.Close()
+
+	for i := 0; i < int(n); i++ {
+		key := make([]byte, rnd.Intn(10)) // TODO Set limit to node.Key max size.?
+		val := make([]byte, rnd.Intn(10))
+
+		if _, err := rnd.Read(key); err != nil {
+			return node.Root{}, fmt.Errorf("failed to create random key")
+		}
+		if _, err := rnd.Read(val); err != nil {
+			return node.Root{}, fmt.Errorf("failed to create random value")
+		}
+
+		if err := tree.Insert(ctx, key, val); err != nil {
+			return node.Root{}, fmt.Errorf("insert entry into tree (%x, %x): %v", key, val, err)
+		}
+	}
+
+	version := 1
+	_, rootHash, err := tree.Commit(ctx, ns, uint64(version))
+	if err != nil {
+		return node.Root{}, fmt.Errorf("commit tree for version %d: %v", version, err)
+	}
+
+	root := node.Root{
+		Namespace: ns,
+		Version:   uint64(version),
+		Type:      node.RootTypeState,
+		Hash:      rootHash,
+	}
+
+	err = ndb.Finalize([]node.Root{root})
+	if err != nil {
+		return node.Root{}, fmt.Errorf("finalize ndb: %v", err)
+	}
+
+	return root, nil
+}
+
+func restoreCheckpoint(ctx context.Context, ndb dbApi.NodeDB, cp *Metadata, fc Creator, root node.Root) error {
+	rs, err := NewRestorer(ndb)
+	if err != nil {
+		return fmt.Errorf("new restorer: %v", err)
+	}
+
+	if err = ndb.StartMultipartInsert(cp.Root.Version); err != nil {
+		return fmt.Errorf("start multipart insert for version %d: %v", cp.Root.Version, err)
+	}
+
+	if err = rs.StartRestore(ctx, cp); err != nil {
+		return fmt.Errorf("start restore (checkpoint: %+v): %v", cp, err)
+	}
+
+	for i := 0; i < len(cp.Chunks); i++ {
+		var cm *ChunkMetadata
+		cm, err = cp.GetChunkMetadata(uint64(i))
+		if err != nil {
+			return fmt.Errorf("get chunk %d metadata: %v", i, err)
+		}
+		var buf bytes.Buffer
+		if err = fc.GetCheckpointChunk(ctx, cm, &buf); err != nil {
+			return fmt.Errorf("get checkpoint chunk %+v: %v", cm, err)
+		}
+		if _, err = rs.RestoreChunk(ctx, uint64(i), &buf); err != nil {
+			return fmt.Errorf("restore chunk (index: %d): %v", i, err)
+		}
+	}
+
+	if err = ndb.Finalize([]node.Root{root}); err != nil {
+		return fmt.Errorf("finalize %+v: %v", root, err)
+	}
+
+	return nil
+}
+
+func ensureEqualEntries(ctx context.Context, ndb1, ndb2 dbApi.NodeDB, root node.Root) error {
+	tree1 := mkvs.NewWithRoot(nil, ndb1, root)
+	defer tree1.Close()
+	it1 := tree1.NewIterator(ctx)
+	defer it1.Close()
+	tree2 := mkvs.NewWithRoot(nil, ndb2, root)
+	defer tree2.Close()
+	it2 := tree2.NewIterator(ctx)
+	defer it2.Close()
+	it1.Rewind()
+	it2.Rewind()
+	for count := 0; it1.Valid(); it1.Next() {
```

Comment:
Note that this doesn't ensure that `ndb2` doesn't have any additional entries. Should ensure that `it2.Next()` reached the end as well.